gtdynamics/universal_robot/sdf.cpp

```diff
@@ -106,11 +113,33 @@ LinkSharedPtr LinkFromSdf(uint8_t id, const sdf::Link &sdf_link) {
   const auto &I = sdf_link.Inertial().Moi();
   inertia << I(0, 0), I(0, 1), I(0, 2), I(1, 0), I(1, 1), I(1, 2), I(2, 0),
       I(2, 1), I(2, 2);
-  const auto wTl = Pose3FromIgnition(sdf_link.Pose());
+
+  /// Call SemanticPose::Resolve so the pose is resolved to the correct frame
+  /// http://sdformat.org/tutorials?tut=pose_frame_semantics&ver=1.7&cat=specification&
+  // Get non-const pose of link in the frame of the joint it is connect to (http://wiki.ros.org/urdf/XML/link).
+  auto raw_pose = sdf_link.RawPose();
+
+  // Update from joint frame to base frame in-place.
+  // Base frame is denoted by "".
+  auto errors = sdf_link.SemanticPose().Resolve(raw_pose, "");
+  // If any errors in the resolution, throw an exception.
+  if (errors.size() > 0) {
+    throw std::runtime_error(errors[0].Message());
+  }
+  // Pose is updated from joint frame to base frame.
+  const auto bTl = Pose3FromIgnition(raw_pose);
   const auto lTcom = Pose3FromIgnition(sdf_link.Inertial().Pose());
+
   return boost::make_shared<Link>(id, sdf_link.Name(),
                                   sdf_link.Inertial().MassMatrix().Mass(),
-                                  inertia, wTl, lTcom);
+                                  inertia, bTl, lTcom);
```

Comment:
So, here it says `bTl`, but the Link docs say `wTl`. I think it's fine, as I think Robot.h should not even know about the "world". But that's for a different PR. But, maybe you want to update the joint to also use the body frame above to be consistent within this file?

Reply:
This was what was confusing me in our call yesterday. Since Robot.h should not know about the world frame, everything should be in the body frame for the robot, is how I understood it, since we could then just update link's pose to be `wTb * bTl` since we have `bTl` for every link.

Reply:
Interestingly, the links are being parsed before the joints (since joints need to know about parent and child links). What this means is that, when parsing the links, we set the link's `wTl` to actually be `bTl` since the robot has no concept of the world frame (at least at the link level). Consequently, the joint's frame being set uses `link->wTl()` which is actually `link->bTl()` since we store `bTl` in the `wTl` property.